tracer/src/Datadog.Trace/Debugger/Symbols/SymbolsUploader.cs

```diff
@@ -0,0 +1,289 @@
+// </copyright>
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Agent.DiscoveryService;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Debugger.Sink;
+using Datadog.Trace.Debugger.Symbols.Model;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.Debugger.Symbols
+{
+    internal class SymbolsUploader : ISymbolsUploader
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SymbolsUploader));
+
+        private readonly JsonSerializerSettings _jsonSerializerSettings;
+        private readonly string _serviceName;
+        private readonly string? _serviceVersion;
+        private readonly string? _environment;
+        private readonly SemaphoreSlim _assemblySemaphore;
+        private readonly SemaphoreSlim _discoveryServiceSemaphore;
+        private readonly HashSet<string> _alreadyProcessed;
+        private readonly long _thresholdInBytes;
+        private readonly CancellationTokenSource _cancellationToken;
+        private readonly IBatchUploadApi _api;
+        private IDiscoveryService? _discoveryService;
+        private byte[]? _payload;
+        private string? _isSymbolUploaderEnabled;
+
+        private SymbolsUploader(IBatchUploadApi api, IDiscoveryService discoveryService, DebuggerSettings settings, ImmutableTracerSettings tracerSettings, string serviceName)
+        {
+            _isSymbolUploaderEnabled = null;
+            _alreadyProcessed = new HashSet<string>();
+            _environment = tracerSettings.EnvironmentInternal;
+            _serviceVersion = tracerSettings.ServiceVersionInternal;
+            _serviceName = serviceName;
+            _discoveryService = discoveryService;
+            _api = api;
+            _assemblySemaphore = new SemaphoreSlim(1);
+            _discoveryServiceSemaphore = new SemaphoreSlim(0);
+            _thresholdInBytes = settings.SymbolBatchSizeInBytes;
+            _cancellationToken = new CancellationTokenSource();
+            _jsonSerializerSettings = new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore };
+            _discoveryService.SubscribeToChanges(ConfigurationChanged);
+        }
+
+        private void ConfigurationChanged(AgentConfiguration configuration)
+        {
+            _isSymbolUploaderEnabled = configuration.SymbolDbEndpoint;
+            if (string.IsNullOrEmpty(Volatile.Read(ref _isSymbolUploaderEnabled)))
+            {
+                Log.Information("Dynamic Instrumentation failed to upload symbols for auto-complete suggestions. This can happen if your datadog-agent version is less than 7.45");
+                return;
+            }
```

Comment:
There's a task waiting on `_discoveryServiceSemaphore` that, in this case, will dangle forever. Isn't it?
I would recommend to release the semaphore, and handle it correctly in the code that is waiting on it (`WaitForDiscoveryServiceAsync`).

Reply:
The question is when we should release it? because we may get to this `ConfigurationChanged` multiple times and we want to wait until `_isSymbolUploaderEnabled`  is not null. WDYT?

Reply:
If that's the case, then the log message is misleading. Isn't it? You log out that the `datadog-agent` version isn't sufficient.

Reply:
We can remove that log or put it in DEBUG level. Will do.

Reply:
Fixed in 6191cf5ad7088cd31a69f1ae170c6c461286578c

Reply:
I don't think that's a blocker, resolving